File: apps/ios/Sources/MouserClient.swift
import SwiftUI

/// App-facing wrapper around the uniffi `MobileClient` (the Rust source/controller in
/// `mouser-ffi`). The trackpad emits **relative** deltas, but the engine forwards
/// absolute motion samples between updates (§7.6), so this integrates deltas into a
/// virtual absolute cursor centred in the engine's coordinate span (mirrors the
/// Android `MouserClient`). All senders are no-ops while disconnected, so the gesture
/// layer can call them unconditionally.
@MainActor
final class MouserClient: ObservableObject {
    @Published private(set) var isConnected = false
    @Published private(set) var status = "Not connected"

    private let inner = MobileClient()

    /// The engine clamps absolute coordinates to the peer's real display; we move a
    /// virtual cursor across a large span and let the peer clamp.
    private static let span: Int32 = 1 << 20
    private var cursorX: Int32 = MouserClient.span / 2
    private var cursorY: Int32 = MouserClient.span / 2

    /// This device's own `device_id` (base32), for display/pairing.
    var deviceId: String { inner.deviceId() }

    /// Dial a peer engine (host/port obtained out-of-band; NWBrowser discovery + a
    /// connect UI are a follow-up, matching Android). Surfaces failures into `status`.
    func connect(host: String, port: UInt16, peerId: String) {
        do {
            try inner.connect(host: host, port: port, peerDeviceIdBase32: peerId)
            cursorX = MouserClient.span / 2
            cursorY = MouserClient.span / 2
            isConnected = true
            status = "Controlling \(host)"
        } catch {
            isConnected = false
            status = "Connect failed: \(error.localizedDescription)"
        }
    }

    func disconnect() {
        inner.disconnect()
        isConnected = false
        status = "Not connected"
    }

    // MARK: - Gesture forwarding

    func move(_ delta: CGSize) {
        guard isConnected else { return }
        cursorX = clampSpan(cursorX + i32(delta.width))
        cursorY = clampSpan(cursorY + i32(delta.height))
        inner.sendPointerMoved(displayId: 0, x: cursorX, y: cursorY)
    }

    func scroll(_ delta: CGSize) {
        guard isConnected else { return }
        inner.sendScroll(dx: i32(delta.width), dy: i32(delta.height))
    }

    /// A full press+release of a pointer button (0=left, 1=right, 2=middle).
    func click(_ button: UInt8) {
        guard isConnected else { return }
        inner.sendButton(button: button, down: true)
        inner.sendButton(button: button, down: false)
    }

    func button(_ button: UInt8, down: Bool) {
        guard isConnected else { return }
        inner.sendButton(button: button, down: down)
    }

    func key(usage: UInt16, down: Bool, mods: UInt16 = 0) {
        guard isConnected else { return }
        inner.sendKey(usage: usage, down: down, mods: mods)
    }

    /// Forward a typed character as a HID press+release, if it maps to a key on a
    /// US layout (`HidKeymap`). Unmappable characters are silently skipped.
    func type(_ character: Character) {
        guard isConnected, let stroke = HidKeymap.stroke(for: character) else { return }
        key(usage: stroke.usage, down: true, mods: stroke.mods)
        key(usage: stroke.usage, down: false, mods: stroke.mods)
    }

    /// Forward a single named key (e.g. Return on submit, Backspace on delete) as a
    /// press+release with no modifiers.
    func tapKey(_ usage: UInt16) {
        guard isConnected else { return }
        key(usage: usage, down: true)
        key(usage: usage, down: false)
    }

    // MARK: - Safe conversions

    /// Round + clamp a CGFloat delta to a sane Int32 (guards NaN/∞ → never traps).
    private func i32(_ value: CGFloat) -> Int32 {
        let rounded = value.rounded()
        guard rounded.isFinite else { return 0 }
        return Int32(min(max(rounded, -100_000), 100_000))
    }

    private func clampSpan(_ value: Int32) -> Int32 {
        min(max(value, 0), MouserClient.span - 1)
    }
}

File: apps/android/app/src/main/kotlin/ai/unlikeother/mouser/companion/MouserClient.kt
package ai.unlikeother.mouser.companion

import android.util.Log
import uniffi.mouser_ffi.MobileClient
import uniffi.mouser_ffi.MobileException
import kotlin.math.roundToInt

/**
 * Thin Kotlin façade over the uniffi-generated [MobileClient] (the shared
 * `mouser-ffi` Rust bridge). It owns the native client and translates the
 * companion's recognised [TouchpadEvent]s into the source/controller FFI calls
 * (`sendPointerMoved` / `sendButton` / `sendScroll` / `sendKey`) — the Android
 * parity of the iOS `MobileClient` wiring.
 *
 * Why a wrapper and not the generated object directly:
 *  - The FFI's `sendPointerMoved` reports an **absolute** cursor sample (the Rust
 *    engine forwards the *motion* between successive samples); the [GestureEngine]
 *    emits **relative** deltas. We integrate the deltas into a virtual position
 *    here so the engine sees a coherent motion stream (§7.6).
 *  - It maps the gesture vocabulary (left/right click, click-drag, scroll) onto
 *    the button/scroll wire calls, and hands typed characters to `sendKey`.
 *  - It keeps the FFI's unsigned types (`UInt`/`UByte`/`UShort`) out of the rest
 *    of the Compose code, and never lets a [MobileException] escape into the UI
 *    thread (every call is best-effort while connected; disconnected events drop).
 *
 * Discovery and pairing are out of scope here (parity with iOS / the FFI doc):
 * [connect] takes an explicit host/port + the peer's base32 `device_id`.
 */
class MouserClient(
    private val client: MobileClient = MobileClient()
) {
    /** Virtual absolute cursor the relative gesture deltas integrate into. Starts
     *  centred in the engine's large virtual span so ordinary motion never clamps. */
    private var cursorX: Int = VIRTUAL_CENTER
    private var cursorY: Int = VIRTUAL_CENTER

    /** This device's own base32 `device_id` (what the peer must pin against). */
    val deviceId: String get() = runCatching { client.deviceId() }.getOrDefault("")

    val isConnected: Boolean get() = runCatching { client.isConnected() }.getOrDefault(false)

    /**
     * Connect to a peer engine at `host:port`, pinning `peerDeviceIdBase32`, and
     * start forwarding input as the source. Returns the failure (if any) rather
     * than throwing, so the UI can surface it without a crash.
     */
    fun connect(host: String, port: Int, peerDeviceIdBase32: String): Result<Unit> = runCatching {
        cursorX = VIRTUAL_CENTER
        cursorY = VIRTUAL_CENTER
        client.connect(host, port.toUShort(), peerDeviceIdBase32)
    }.onFailure { Log.w(TAG, "connect failed", it) }

    /** Tear down the session (idempotent). */
    fun disconnect() {
        runCatching { client.disconnect() }.onFailure { Log.w(TAG, "disconnect failed", it) }
    }

    /**
     * Route one recognised gesture event to the wire. A no-op while disconnected.
     * `Magnify`/`Rotate` have no source-input wire form yet (the FFI exposes only
     * pointer/button/scroll/key), so they are dropped here — they remain local
     * haptic/readout affordances until the protocol carries them.
     */
    fun onEvent(event: TouchpadEvent) {
        if (!isConnected) return
        when (event) {
            is TouchpadEvent.Move -> moveBy(event.dx, event.dy)
            is TouchpadEvent.Scroll -> send { sendScroll(event.dx.roundToInt(), event.dy.roundToInt()) }
            TouchpadEvent.LeftClick -> click(BUTTON_LEFT)
            TouchpadEvent.RightClick -> click(BUTTON_RIGHT)
            TouchpadEvent.DragStart -> send { sendButton(BUTTON_LEFT.toUByte(), true) }
            TouchpadEvent.DragEnd -> send { sendButton(BUTTON_LEFT.toUByte(), false) }
            // Local-only feedback; no source wire representation yet.
            TouchpadEvent.ScrollDetent,
            is TouchpadEvent.Magnify,
            is TouchpadEvent.Rotate -> Unit
        }
    }

    /**
     * Send a typed character as a press+release of its HID usage. Only the subset
     * the capture field produces (ASCII letters/digits/space/return) is mapped;
     * unmapped characters are ignored. Modifier handling (shift for capitals) is a
     * follow-up — the engine lower-cases to the base usage today.
     */
    fun sendCharacter(ch: Char) {
        val usage = hidUsageFor(ch) ?: return
        send {
            sendKey(usage, true, NO_MODS)
            sendKey(usage, false, NO_MODS)
        }
    }

    private fun moveBy(dx: Float, dy: Float) {
        cursorX = (cursorX + dx.roundToInt()).coerceIn(0, VIRTUAL_SPAN)
        cursorY = (cursorY + dy.roundToInt()).coerceIn(0, VIRTUAL_SPAN)
        send { sendPointerMoved(0u, cursorX, cursorY) }
    }

    private fun click(button: Int) = send {
        sendButton(button.toUByte(), true)
        sendButton(button.toUByte(), false)
    }

    /** Run an FFI block, swallowing a [MobileException] so a stray gesture during
     *  teardown can never crash the UI thread. */
    private inline fun send(block: MobileClient.() -> Unit) {
        try {
            client.block()
        } catch (e: MobileException) {
            Log.w(TAG, "send dropped", e)
        }
    }

    private companion object {
        const val TAG = "MouserClient"

        // §7.5 pointer button indices.
        const val BUTTON_LEFT = 0
        const val BUTTON_RIGHT = 1

        val NO_MODS: UShort = 0u

        // Mirrors mouser-ffi VIRTUAL_SPAN (1 << 20); the deltas integrate inside it.
        const val VIRTUAL_SPAN = 1 shl 20
        const val VIRTUAL_CENTER = VIRTUAL_SPAN / 2

        /** Minimal USB HID Usage Page 0x07 map for the capture field's ASCII subset. */
        fun hidUsageFor(ch: Char): UShort? = when (ch) {
            in 'a'..'z' -> (0x04 + (ch - 'a')).toUShort()
            in 'A'..'Z' -> (0x04 + (ch - 'A')).toUShort()
            in '1'..'9' -> (0x1E + (ch - '1')).toUShort()
            '0' -> 0x27u
            ' ' -> 0x2Cu
            '\n' -> 0x28u // Return/Enter
            else -> null
        }
    }
}
